projects/vdk-plugins/vdk-meta-jobs/tests/test_meta_job.py

```diff
@@ -303,3 +303,55 @@ def test_meta_job_circular_dependency(httpserver: PluginHTTPServer):
         # no other request should be tried as the meta job fails
         assert isinstance(result.exception, UserCodeError)
         assert len(httpserver.log) == 0
+
+
+def test_meta_job_concurrent_running_jobs_limit(httpserver: PluginHTTPServer):
+    jobs = [("job" + str(i), [200], "succeeded") for i in range(1, 8)]
+    api_url = _prepare(httpserver, jobs)
+
+    with mock.patch.dict(
+        os.environ,
+        {
+            "VDK_CONTROL_SERVICE_REST_API_URL": api_url,
+            "VDK_META_JOBS_MAX_CONCURRENT_RUNNING_JOBS": "5",
+            "VDK_META_JOBS_DELAYED_JOBS_MIN_DELAY_SECONDS": "1",
+            "VDK_META_JOBS_DELAYED_JOBS_RANDOMIZED_ADDED_DELAY_SECONDS": "1",
+        },
+    ):
+        # CliEntryBasedTestRunner (provided by vdk-test-utils) gives a way to simulate vdk command
+        # and mock large parts of it - e.g passed our own plugins
+        runner = CliEntryBasedTestRunner(plugin_entry)
+
+        result: Result = runner.invoke(
+            ["run", jobs_path_from_caller_directory("meta-job-exceed-limit")]
+        )
+
+        expected_max_running_jobs = int(
+            os.getenv("VDK_META_JOBS_MAX_CONCURRENT_RUNNING_JOBS", "5")
+        )
+        post_reqs = 0
+        # keep track of the number of running jobs at any given time
+        running_jobs = 0
+        # track the latest request in order to reset the number of running jobs when such finish
+        latest_req = None
+        for request, response in httpserver.log:
+            if "executions" in request.path and request.method == "POST":
+                running_jobs += 1
+                post_reqs += 1
+                assert (
+                    running_jobs <= expected_max_running_jobs
+                )  # assert that max concurrent running jobs is not
+                # exceeded
+            if (
+                request.method == latest_req
+                and "executions/job" in request.path
+                and response.status < "300"
+            ):
+                running_jobs = 0  # back-to-back GET reqs means the jobs have been finalized successfully
```

Comment:
When testing, it's important to focus on testing what the software is supposed to do, instead of how it's built.  
* If the implementation changes, the tests will need to be updated. This is especially bad for refactoring changes. Refactoring is the process of making changes to the code to improve it  without changing its behavior.
* Tests that depend on specific implementation details are limited in scope and are likely to catch less potential bugs

Reply:
So I think it would be better instead to keep a set of jobs (the key in the set would be the job name) when a job is started we add to a set, and if we get status finished response we remove the job from the set . Wouldn't this be more robust ?

Reply:
Yep, it would be. I was way too excited to have a working test and got carried away... Will change according to your feedback.